feat(eebus): add EEBUS module with LPC use case

mlitre opened the pull request:
## Describe your changes

Adds an `EEBUS` module (`modules/EnergyManagement/EEBUS`) that bridges EVerest to an EEBUS-capable energy manager and implements the **Limitation of Power Consumption (LPC)** use case.

The module talks to the [eebus-grpc-api](https://github.com/EVerest/eebus-grpc-api) Go sidecar over gRPC. It can optionally manage the sidecar's lifecycle itself (`manage_eebus_grpc_api_binary`): start it, restart it on crash with a configurable delay, and generate its TLS certificate/key on first run.

**LPC use case.** `lpc::LpcUseCaseHandler` implements the LPC state machine (`Init`, `Limited`, `Unlimited/controlled`, `Unlimited/autonomous`, `Failsafe`) with an injectable clock for deterministic unit testing:

- Heartbeat supervision with a 120 s timeout driving Failsafe entry [LPC-911/912] and the independent Failsafe exit conditions [LPC-916, LPC-921, LPC-922].
- Failsafe consumption limit applied already in `Init` [LPC-901/1] and re-applied on Failsafe entry; cached EG state is discarded on entry to `Failsafe`/`Unlimited/autonomous` so stale limits cannot drive transitions [LPC-918/919/920].
- Validation of EG writes: negative Active Power Consumption Limits are NACKed [LPC-001] and Failsafe Duration Minimum writes are constrained to the [2 h, 24 h] range [LPC-022].
- Received limits are translated into `ExternalLimits` schedules and published through the `eebus_energy_sink` requirement (`external_energy_limits` interface), typically consumed by `EnergyManager`.

**Trust management.** A configurable EMS SKI allowlist (`eebus_ems_ski_allowlist`) is registered with the sidecar at startup; the module subscribes to discovery events and re-registers allowlisted EGs that appear at runtime. An opt-in `accept_unknown_ems` flag (default `false`, security-sensitive) auto-trusts unknown EGs for isolated networks.

**Module structure.** Shared infrastructure (connection handling, config validation, gRPC event readers) lives at the module root; use-case-specific code lives in a per-use-case subdirectory (`lpc/`) so future use cases (LPP, MPC, …) can be added alongside. The code follows the C++ coding guidelines introduced in #2017 (`m_` member prefixes, `#pragma once`, doxygen comments, co-located headers).

**Testing.**

- C++ unit tests (gtest, run via ctest): LPC state machine incl. heartbeat/failsafe timing and multi-EG SKI filtering with a fake clock, config validator incl. allowlist parsing, and `UseCaseEventReader` shutdown behavior.
- Python integration suite under `tests/eebus_tests` with mock gRPC control/LPC servicers (`./run-tests.sh eebus`), covering load limits with/without duration, failsafe values, heartbeat, allowlist and discovery flows.
- Module documentation under `modules/EnergyManagement/EEBUS/docs` (architecture, state machine, code flow, class diagrams).

## Issue ticket number and link

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have made corresponding changes to the documentation
- [x] I read the [contribution documentation](https://everest.github.io/nightly/project/contributing.html) and made sure that my changes meet its requirements

andistorm committed:
feat(EEBus): Implement EEBus LPC Use Case

This is the first implementation of EEBus and the LPC use case. The
EEBUS module requires an energy sink to limit the energy flow request
going to the energy manager.

Co-authored-by: Andreas Heinrich <andreas.heinrich@rwth-aachen.de>
Signed-off-by: Andreas Heinrich <andreas.heinrich@rwth-aachen.de>
Signed-off-by: Martin Litre <mnlitre@gmail.com>

andistorm commented on .ci/build-kit/scripts/create_ocpp_tests_image.sh:
```diff
@@ -8,7 +8,7 @@ if [ $retVal -ne 0 ]; then
 fi
 
-pip install --break-system-packages \
+python3 -m pip install --break-system-packages \
```
Could be in a dedicated PR "Improce CI scripts" or so, but not required from my side

andistorm commented on .ci/build-kit/scripts/create_ocpp_tests_image.sh:
```diff
@@ -20,6 +20,6 @@ if [ $retVal -ne 0 ]; then
 fi
 
-pip install --break-system-packages -r tests/ocpp_tests/requirements.txt
+python3 -m pip install --break-system-packages -r tests/ocpp_tests/requirements.txt
```
Could be in a dedicated PR "Improce CI scripts" or so, but not required from my side

andistorm commented on .ci/e2e/scripts/run_ocpp_tests.sh:
```diff
@@ -6,7 +6,7 @@ PARALLEL_TESTS=$(nproc)
 echo "Running $PARALLEL_TESTS ocpp tests in parallel"
 
-pytest \
+python3 -m pytest \
```
Could be in a dedicated PR "Improce CI scripts" or so, but not required from my side

andistorm commented on .github/workflows/build_and_test.yaml:
```diff
@@ -30,6 +30,7 @@ jobs:
       do_not_run_coverage_badge_creation: true
       run_install_wheels: true
       run_integration_tests: true
+      build_kit_base_image_tag: feature-eebus
```
This should be updated to a proper tag before merging

andistorm commented on modules/EnergyManagement/EEBUS/manifest.yaml:
```diff
@@ -0,0 +1,71 @@
+    type: string
+    required: true
+  certificate_path:
+    description: Path to the certificate file used by eebus go client. If relative will be prefixed with everest prefix + etc/everest/certs. Otherwise absolute file path is used.
```
```suggestion
    description: Path to the certificate file used by eebus go client. If relative will be prefixed with everest prefix + etc/everest/certs. Otherwise absolute file path is used, required/used when manage_eebus_grpc_api_binary=true
```

andistorm commented on modules/EnergyManagement/EEBUS/manifest.yaml:
```diff
@@ -0,0 +1,71 @@
+    description: Serial number for the configuration of the control service
+    type: string
+    required: true
+  failsafe_control_limit:
```
```suggestion
  failsafe_control_limit_W:
```
Adding the unit as suffix could help to avoid mistakes in code

andistorm commented on tests/eebus_tests/grpc_server/control_service_server.py:
```diff
@@ -0,0 +1,41 @@
+        await self._server.start()
+
+    async def stop(self):
+        await self._server.stop(0)
```
```suggestion
        await self._server.stop(0)

```

andistorm commented on tests/eebus_tests/grpc_servicer/cs_lpc_control_servicer.py:
```diff
@@ -0,0 +1,86 @@
+        return await default_command_func(self, request, context, "ConsumptionNominalMax")
+
+    async def SetConsumptionNominalMax(self, request, context):
+        return await default_command_func(self, request, context, "SetConsumptionNominalMax")
```
```suggestion
        return await default_command_func(self, request, context, "SetConsumptionNominalMax")

```

andistorm commented on tests/eebus_tests/helpers/async_helpers.py:
```diff
@@ -0,0 +1,13 @@
+    try:
+        return await asyncio.wait_for(q.get(), timeout=timeout)
+    except asyncio.TimeoutError:
+        raise TimeoutError(f"async_get timed out after {timeout} seconds")
```
```suggestion
        raise TimeoutError(f"async_get timed out after {timeout} seconds")

```

mlitre commented on modules/EnergyManagement/EEBUS/doc.rst:
```diff
@@ -0,0 +1,328 @@
+   LpcUseCaseHandler ..> eebus.EEBusCallbacks
+   UseCaseEventReader ..> LpcUseCaseHandler : event_callback
+
+   @enduml
```
Maybe I am using a newer version of plantuml which allows me to render new things

andistorm committed:
Remove deprecated generous ci workflow call

Signed-off-by: Andreas Heinrich <andreas.heinrich@rwth-aachen.de>
Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
fix(eebus): address issues in connection lifecycle / error handling

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
improvements to object handling

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
fix(EEBus): Update the UseCaseSupportUpdate for proper connection

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
add doc on running tests

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
fix format

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
Update python dependencies

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
fix eebus module shutdown mechanism

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
feat(eebus): Make it possible to generate certs on startup

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
fix(eebus): Improve logging for ebbus grpc binary

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
harden state transition + remove unused param

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
fix eebus config example

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
fix create_ocpp_tests_image script

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
feat(eebus-tests): enable parallel execution and add CI suite

- Add EebusPortStrategy that dynamically allocates grpc_port and
  eebus_service_port, replacing hardcoded 50051/4715 that prevented
  parallel test execution.
- Add async_wait_for helper that drains intermediate queue events
  until the expected value arrives, replacing tight timeout windows
  and inline asyncio.sleep calls that caused flaky failures under
  CPU contention.
- Add 'eebus' suite to the unified test runner (run-tests.sh).
- Add EEBUS test step to the integration CI workflow.

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
fix(eebus-tests): guard duplicate setup_grpc() call

With -DGRPC_EDM=ON, setup_grpc() does add_library(gRPC::grpc++ ALIAS
grpc++). The EEBUS feature calls the macro from both
modules/EnergyManagement/EEBUS/grpc_libs/ and tests/eebus_tests/grpc_libs/,
so the second call errors with "ALIAS target already exists". Guard on
NOT TARGET gRPC::grpc++ and set _GRPC_SETUP_DONE explicitly so the
following setup_grpc_generator() call still succeeds.

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
docs(eebus): fix LPC-ID citations in state-machine diagram

Four citations referenced non-existent or wrong IDs per
EEBus UC TS V1.0.0 §2.3.3:

- Limited → Unl/ctrl: [LPC-914/915] → [LPC-908/909] (LPC-915 doesn't exist)
- Limited → Failsafe: [LPC-911/912] → [LPC-912]
- Unl/ctrl → Limited: [LPC-913] → [LPC-910]
- Failsafe → Unl/ctrl: [LPC-916/918] → [LPC-918/920]

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
feat(eebus): NACK negative APCL writes [LPC-001]

EEBus UC TS V1.0.0 line 313: "A limit lower than 0W SHALL be rejected."
approve_pending_writes() previously ACKed everything, so a negative value
would propagate a nonsense negative total_power_W through ExternalLimits
into EnergyManager. Now any pending limit with value < 0 is NACKed via
ApproveOrDenyConsumptionLimit with reason
"LPC-001: Active Power Consumption Limit must be >= 0 W".

The limit_value_is_valid predicate is a public static — a pure validator
natural to the class's API. Tests call it directly, no friendships.

Adds 4 C++ validator tests and a Python integration test asserting the
NACK end-to-end.

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
feat(eebus): reject negative FailsafeConsumptionActivePowerLimit [LPC-001]

EEBus UC TS V1.0.0 line 404: the Failsafe Consumption Active Power Limit
SHALL be ≥ 0. The handler previously wrote whatever the sidecar reported,
so a negative value would become the applied power cap in Init/Failsafe
states and propagate through ExternalLimits.

As with FailsafeDurationMinimum, no pending/approve flow exists for this
datapoint — defensive cache preservation only. failsafe_limit_is_valid
is a public static predicate matching the pattern of the other two
validators.

Adds 3 C++ validator tests and a Python integration test asserting that
Failsafe still applies the 4200W config default after a -1000W write.

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
feat(eebus): enforce FailsafeDurationMinimum [2h, 24h] range [LPC-022]

EEBus UC TS V1.0.0 §2.6.2.1: the CS SHALL accept EG writes only in
[2h, 24h] for the Failsafe Duration Minimum. Previously the handler
stored whatever the sidecar reported, so a 5s value would cause
premature Failsafe exit and violate [LPC-922].

Models the spec bounds as two constexpr values in the anonymous
namespace:
- LPC_VENDOR_CONFIG_MIN (2h) — [LPC-022/1] vendor-config floor
- LPC_CS_MAXIMUM (24h) — [LPC-022/4] CS's max accepted write

failsafe_duration_is_valid(duration) is a public static predicate
gating the runtime check in handle_data_update_failsafe_duration_minimum().
Out-of-range writes are logged as a warning and dropped; the cached
value is kept. No pending/approve gRPC flow exists for this datapoint.

The existing Python test_update_failsafe_duration_minimum used a 5s
value now out-of-range; renamed and inverted to
test_ignores_out_of_range_failsafe_duration. C++ adds 6 validator tests.

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
build(eebus): bump eebus-grpc fork and regenerate control_service stubs

Point dependencies.yaml at the Pionix fork (mlitre/eebus-grpc) at the
commit that adds the SubscribeDiscoveryEvents server-streaming RPC and
the DiscoveryEvent message, plus a guard-widening on RegisterRemoteSki
so it can be called both before and after StartService.

Regenerates the C++ and Python stubs under
modules/EnergyManagement/EEBUS/grpc_libs/generated/control_service/ and
tests/eebus_tests/grpc_libs/generated/control_service/ to pick up the
new RPC. These are pure protoc output; the feature implementation lives
in the follow-up module-side commit.

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
feat(eebus): add EMS SKI allowlist and runtime discovery trust

Replaces the single eebus_ems_ski config with a two-part trust model:

- eebus_ems_ski_allowlist (comma-separated 40-hex SKIs, whitespace
  tolerated) is deduped and normalized into an effective allowlist at
  init and registered with the sidecar one-by-one before StartService.
- accept_unknown_ems (boolean, default false) gates runtime auto-trust
  of any SKI the sidecar reports as newly discovered but unknown.
  Security-sensitive — documented as safe only on isolated networks.

The module subscribes to the new SubscribeDiscoveryEvents stream via a
DiscoveryEventReader (mirrors UseCaseEventReader's ClientReadReactor
pattern with destructor-join guarantees) and classifies every
DISCOVERED event into one of four actions: already-trusted no-op,
allowlisted re-register, unknown auto-accept (when the flag is true),
or unknown-ignored.

Adds single-EG semantics in LpcUseCaseHandler so events from multiple
discovered peers don't interleave state transitions: the first data-
carrying event's SKI is recorded as the active EG; subsequent events
from other SKIs are dropped. The active EG is cleared on entry to
Failsafe (lost contact) or UnlimitedAutonomous (reachable from Init
via 120s timeout [LPC-906] with an EG nominally connected — exit per
[LPC-918/919/920] needs heartbeat + limit from some EG, so a fresh
connection must be accepted).

UseCaseEventReader's callback is widened from UseCaseEvent to
SubscribeUseCaseEventsResponse so the LPC handler can see remote_ski.

ConfigValidator uses a libocpp-style split_string helper (copied, not
linked) rather than taking a runtime dep on libocpp.

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
test(eebus): add C++ unit tests for SKI allowlist and active-EG filtering

config_validator_tests (new target): 10 cases covering CSV parse,
lowercase + whitespace normalization, dedupe, empty-allowlist +
accept_unknown combinations, and format rejection (non-hex / short /
long entries). Default-constructs a Conf with manage_eebus_grpc_api_binary
=false to skip the filesystem validators.

lpc_state_machine_tests (extended): 5 new cases asserting the single-EG
semantics in LpcUseCaseHandler — first data event records the active
EG, heartbeats from a second SKI are ignored (and the primary EG's
timeout still fires Failsafe), Failsafe + UnlimitedAutonomous entries
clear the active EG so a different EG can take over, and non-data
events (UseCaseSupportUpdate) from a non-active SKI are filtered.

Introduces static const SKI_A/SKI_B constants at test-file scope so
multi-EG scenarios read clearly rather than inlining std::string(40,
'a') everywhere.

Signed-off-by: Martin Litre <mnlitre@gmail.com>

mlitre committed:
test(eebus-tests): add integration coverage for allowlist and discovery

Mock-gRPC (TestEEBUSAllowlistDiscovery — 7 new tests):
- Allowlist with 3 SKIs registers 3 times at boot.
- DISCOVERED event for an allowlisted SKI with is_trusted=false
  triggers a runtime RegisterRemoteSki.
- DISCOVERED event with is_trusted=true is a no-op.
- accept_unknown_ems=true auto-registers unknown SKIs; =false ignores.
- Active-EG connection: events from a second SKI are dropped while
  the first is connected.
- Failover through Failsafe: heartbeat timeout clears the active EG
  and a second SKI takes over.

Real-stack (TestEEBUSReferenceAutoDiscovery — new):
- Empty allowlist + accept_unknown_ems=true paired with the eebus-go
  reference controlbox. Exercises the full mDNS → DiscoveryEvent →
  runtime RegisterRemoteSki → SHIP pairing → limit delivery path that
  the mock tests intentionally don't cover.

Mock infra:
- control_service_servicer adds SubscribeDiscoveryEvents (unbounded
  response queue, same shape as SubscribeUseCaseEvents) and makes
  RegisterRemoteSki's queues unbounded so tests can drive multi-SKI
  allowlists and runtime discovery calls.
- EebusModuleConfigStrategy (conftest) applies a dict of
  config_module overrides via @pytest.mark.everest_config_adaptions,
  keeping per-test config in the test body rather than in separate
  yaml files.
- perform_eebus_handshake takes expected_register_ski_count so callers
  can drain the boot-time registers before asserting on the module's
  subsequent behavior.

Migrates the shipped test configs from eebus_ems_ski to
eebus_ems_ski_allowlist to match the new schema.

Signed-off-by: Martin Litre <mnlitre@gmail.com>

andistorm reviewed (dismissed):
🐧

mlitre committed:
feat(eebus): add EEBUS module with LPC use case

Signed-off-by: Martin Litre <mnlitre@gmail.com>
